src/uu/head/src/head.rs

```diff
@@ -1,240 +1,557 @@
-                    break;
-                }
+    let mut ringbuf = vec![0u8; n];
 
-                // If this is the last number
-                if pos == len - 1 {
-                    options.remove(a);
-                    let number: Option<usize> =
-                        from_utf8(&current[1..len]).unwrap().parse::<usize>().ok();
-                    return (options, Some(number.unwrap()));
-                }
+    // first we fill the ring buffer
+    if let Err(e) = input.read_exact(&mut ringbuf) {
+        if e.kind() == ErrorKind::UnexpectedEof {
+            return Ok(());
+        } else {
+            return Err(e);
+        }
+    }
+    let mut buffer = [0u8; BUF_SIZE];
+    loop {
+        let read = loop {
+            match input.read(&mut buffer) {
+                Ok(n) => break n,
+                Err(e) => match e.kind() {
+                    ErrorKind::Interrupted => {}
+                    _ => return Err(e),
+                },
+            }
+        };
+        if read == 0 {
+            return Ok(());
+        } else if read >= n {
+            stdout.write_all(&ringbuf)?;
+            stdout.write_all(&buffer[..read - n])?;
+            for i in 0..n {
+                ringbuf[i] = buffer[read - n + i];
             }
+        } else {
+            stdout.write_all(&ringbuf[..read])?;
+            for i in 0..n - read {
+                ringbuf[i] = ringbuf[read + i];
+            }
+            ringbuf[n - read..].copy_from_slice(&buffer[..read]);
         }
-
-        a += 1;
     }
+}
 
-    (options, None)
+fn rbuf_but_last_n_lines(
+    input: &mut impl std::io::BufRead,
+    n: usize,
+    zero: bool,
+) -> std::io::Result<()> {
+    if n == 0 {
+        //prints everything
+        return rbuf_n_bytes(input, std::usize::MAX);
+    }
+    let mut ringbuf = vec![Vec::new(); n];
+    let stdout = std::io::stdout();
+    let mut stdout = stdout.lock();
+    let mut line = Vec::new();
+    let mut lines = 0usize;
+    split::walk_lines(input, zero, |e| match e {
+        split::Event::Data(dat) => {
+            line.extend_from_slice(dat);
+            Ok(true)
+        }
+        split::Event::Line => {
+            if lines < n {
+                ringbuf[lines] = std::mem::replace(&mut line, Vec::new());
+                lines += 1;
+            } else {
+                stdout.write_all(&ringbuf[0])?;
+                ringbuf.rotate_left(1);
+                ringbuf[n - 1] = std::mem::replace(&mut line, Vec::new());
+            }
+            Ok(true)
+        }
+    })
 }
 
-// TODO: handle errors on read
-fn head<T: Read>(reader: &mut BufReader<T>, settings: &Settings) -> bool {
-    match settings.mode {
-        FilterMode::Bytes(count) => {
-            for byte in reader.bytes().take(count) {
-                print!("{}", byte.unwrap() as char);
+fn head_backwards_file(input: &mut std::fs::File, options: &HeadOptions) -> std::io::Result<()> {
```

Comment:
Looking at the number of warnings, it doesn't seem tested ?

Reply:
Oh, yeah. Looking at the tests, all inputs are piped.

Reply:
> Looking at the number of warnings, it doesn't seem tested ?

btw, where do you see these warnings?

Reply:
Here: https://github.com/uutils/coreutils/pull/1911/files 
when the coverage run are completed

You can also what it here:
https://codecov.io/gh/uutils/coreutils/commit/4ee61a9e033609b16e233e6e61ce34ab63b4ee58/

Reply:
and here: https://codecov.io/gh/uutils/coreutils/src/3b8dcbf0cfe99657ce7e9b59c8d4f73db33ff5b6/src/uu/head/src/head.rs

Reply:
oh nice, didn't know that